### PR TITLE
Expose dots which haven't been matched to a paper as markers

### PR DESCRIPTION
**Pull request by paulsonnentag**
I wanted to have markers which are tracked independently of the papers, similar to issue #42. I thought the fastest implementation would be to reuse the existing circle detection of the paper recognition. Any point which isn't part a corner shape is exposed as a marker on each paper object

```JavaScript
const papers = await paper.get('papers');
const number = await paper.get('number');

// each paper has list of markers
papers[number].markers
```

I've build a little curling game with it:

img_20180414_204202
https://www.youtube.com/watch?v=0wXpnWLRP1Y

The only drawback I've noticed is that if a corner of a paper is partially covered the remaining dots are no longer part of a shape and will be recognized as markers. In my demo, I've dealt with this by using bigger circles for the markers and filter for the size in my paper program. The implementation isn't complete right now; I would also like to add the information which of the markers are on the paper and their relative position to the paper. I'm interested to hear your feedback.

### CODE REVIEW

**Comment by shaunlebron** [comment]
> I thought the fastest implementation would be to reuse the existing circle detection of the paper recognition.

Cool idea, when ready we can look at the square detection code [here](https://github.com/opencv/opencv/blob/master/samples/cpp/squares.cpp)

> I would also like to add the information which of the markers are on the paper and their relative position to the paper.

👍

File: client/camera/detectPrograms.js [hunk 1]
```diff
@@ -398,6 +403,19 @@ export default function detectPrograms({ config, videoCapture, dataToRemember, d
...
+    .filter(({ matchedShape }) => !matchedShape)
+    .map(({ colorIndex, avgColor, pt, size }) => {
+      const { x, y } = projectPointToUnitSquare(pt, videoMat, config.knobPoints);
+      return { size, position: { x, y }, avgColor, colorIndex };
+    });
+
+  // add markers to programs
+  programsToRender.forEach(programToRender => {
+    programToRender.markers = markers;
+  });
```
Comment (janpaul123) [inline client/camera/detectPrograms.js]: Exposing the markers as data on every paper seems a bit strange to me. Maybe allow access to them via `paper.get('markers');`? Or only expose the markers that are contained within a paper on the paper (similarly to how it works on DL)?

File: client/camera/detectPrograms.js [hunk 2]
```diff
@@ -398,6 +403,19 @@ export default function detectPrograms({ config, videoCapture, dataToRemember, d
     }
   });
 
+  // all points which haven't been matched to a shape are added as markers to the paper data
+  const markers = keyPoints
+    .filter(({ matchedShape }) => !matchedShape)
+    .map(({ colorIndex, avgColor, pt, size }) => {
+      const { x, y } = projectPointToUnitSquare(pt, videoMat, config.knobPoints);
+      return { size, position: { x, y }, avgColor, colorIndex };
```
Comment (janpaul123) [inline client/camera/detectPrograms.js]: `colorIndex` might not be super meaningful to programs. How about omitting that and changing `avgColor` to simply `color`?

Comment (janpaul123) [reply client/camera/detectPrograms.js]: Ah I see how you used the `colorIndex` in the curling game and how that was convenient. Hm, I could get on board with that. Though might be nicer then to convert it to a colour name ("red", "green", etc)?

File: client/camera/detectPrograms.js (hunk 1 above)
Comment (sgwilym) [reply client/camera/detectPrograms.js]: Maybe having both would be valuable?

- `myPaper.markers` exposes markers on `myPaper`
- `paper.get('markers')` exposes all markers that are _not_ on any piece of paper.

File: client/camera/detectPrograms.js (hunk 2 above)
Comment (paulsonnentag) [reply client/camera/detectPrograms.js]: Agree with you, I'll change that.

File: client/camera/detectPrograms.js (hunk 1 above)
Comment (paulsonnentag) [reply client/camera/detectPrograms.js]: I think the most flexible way would be to receive all markers with `paper.get('markers')`. I thought about having both APIs like @sgwilym suggested, but then it get's complicated for some case for example If you want to get all markers. Instead we could add a `paperId` property and `positionOnPaper` property if the marker is on a paper. 

Marker Object:

```JavaScript
{ 
  position: {x, y},  // global position
  color: [r, g, b]
  colorName // "red", "green", "blue" or "black"

  // these properties will be null if the marker is not on a paper
  paperId,
  positionOnPaper: {x, y} // position converted to coordinate system of the paper
}
```

Comment (janpaul123) [reply client/camera/detectPrograms.js]: I like that idea a lot!

**Comment by shaunlebron** [comment]
Great start!  We played with this, and drew circles over all the markers:

markers

Looks like the markers aren't attached to papers yet, so I'll remove that from the docs.

**Comment by paulsonnentag** [comment]
Yes, I've started working on that but haven't finished it yet.

**Pull request merged**